Add dynamic feature support

princeparadoxes opened the pull request.

Commit:
Add dynamic feature support

princeparadoxes commented on configuration/src/main/kotlin/com/malinskiy/marathon/config/vendor/android/TimeoutConfiguration.kt:
```diff
@@ -10,7 +10,7 @@ data class TimeoutConfiguration(
     var pullFile: Duration = Duration.ofSeconds(30),
     var uninstall: Duration = shell,
-    var install: Duration = shell,
+    var install: Duration = Duration.ofSeconds(60),
```
Пришлось увеличить timeout, так как split apk устанавливаются достаточно долго

Malinskiy replied:
Без проблем

princeparadoxes commented on vendor/vendor-android/adam/src/main/kotlin/com/malinskiy/adam/request/pkg/InstallSplitPackageRequest.kt:
```diff
@@ -0,0 +1,93 @@
+                listOf(pkg),
+                supportedFeatures,
+                reinstall,
+                extraArgs
```
В оригинале в Adam [такая строчка ](https://github.com/Malinskiy/adam/blob/0e0ba80a47ced71992c98ec0172d7d1ff3637db9/adam/src/main/kotlin/com/malinskiy/adam/request/pkg/InstallSplitPackageRequest.kt#L71)
`extraArgs + "-S${totalSize}"`
С ней выпадает такая ошибка
```
java.lang.IllegalArgumentException: No argument expected after "-g -S257177037"
        at android.os.BasicShellCommandHandler.getNextOption(BasicShellCommandHandler.java:214)
        at com.android.server.pm.PackageManagerShellCommand.makeInstallParams(PackageManagerShellCommand.java:2765)
        at com.android.server.pm.PackageManagerShellCommand.runInstallCreate(PackageManagerShellCommand.java:1425)
        at com.android.server.pm.PackageManagerShellCommand.onCommand(PackageManagerShellCommand.java:197)
        at android.os.BasicShellCommandHandler.exec(BasicShellCommandHandler.java:98)
        at android.os.ShellCommand.exec(ShellCommand.java:44)
        at com.android.server.pm.PackageManagerService.onShellCommand(PackageManagerService.java:22097)
        at android.os.Binder.shellCommand(Binder.java:929)
        at android.os.Binder.onTransact(Binder.java:813)
        at android.content.pm.IPackageManager$Stub.onTransact(IPackageManager.java:4621)
        at com.android.server.pm.PackageManagerService.onTransact(PackageManagerService.java:4639)
        at android.os.Binder.execTransactInternal(Binder.java:1159)
        at android.os.Binder.execTransact(Binder.java:1123)
```
Попробовал изменить порядок вот так
`listOf("-S${totalSize}") + extraArgs`
но тогда опять же ошибка
```
java.lang.NumberFormatException: For input string: "257177037 -g"
        at java.lang.Long.parseLong(Long.java:594)
        at java.lang.Long.parseLong(Long.java:636)
        at com.android.server.pm.PackageManagerShellCommand.makeInstallParams(PackageManagerShellCommand.java:2817)
        at com.android.server.pm.PackageManagerShellCommand.runInstallCreate(PackageManagerShellCommand.java:1425)
        at com.android.server.pm.PackageManagerShellCommand.onCommand(PackageManagerShellCommand.java:197)
        at android.os.BasicShellCommandHandler.exec(BasicShellCommandHandler.java:98)
        at android.os.ShellCommand.exec(ShellCommand.java:44)
        at com.android.server.pm.PackageManagerService.onShellCommand(PackageManagerService.java:22097)
        at android.os.Binder.shellCommand(Binder.java:929)
        at android.os.Binder.onTransact(Binder.java:813)
        at android.content.pm.IPackageManager$Stub.onTransact(IPackageManager.java:4621)
        at com.android.server.pm.PackageManagerService.onTransact(PackageManagerService.java:4639)
        at android.os.Binder.execTransactInternal(Binder.java:1159)
        at android.os.Binder.execTransact(Binder.java:1123)
```
В итоге вообще убрал параметр -S. Не совсем понял для чего этот параметр вообще нужен.

princeparadoxes replied:
Пока вижу вариант в Adam убрать добавление этого параметра. Это норм?

Malinskiy replied:
Там нужно указывать полный размер пакета до передачи насколько я помню, иначе девайс не может детектировать конец стрима. Может пробел просто отсутствует?

princeparadoxes commented on vendor/vendor-android/base/src/test/kotlin/com/malinskiy/marathon/android/AndroidAppInstallerTest.kt:
```diff
@@ -71,6 +71,46 @@ class AndroidAppInstallerTest {
+                    shell("pm list packages", "")
+                    installApk(temp, "/data/local/tmp/app-debug.apk", "511", "122fc3b5d69b262db9b84dfc00e8f1d4", "-r")
+
+                    // TODO
```
Не до конца понимаю как написать нужный мне тест.

princeparadoxes replied:
Ещё когда вызываю
`./gradlew clean test jacocoTestReport integrationTest`
то новый мой тест как будто игнорируется

Malinskiy replied:
там возможно нужно стаб сервера еще подкрутить: он делается по мере необходимости, а то весь адб сервер иначе надо писать. я посмотрю что нужно еще

princeparadoxes commented on vendor/vendor-android/adam/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt:
```diff
@@ -295,6 +298,22 @@ class AdamAndroidDevice(
+                ), serial = adbSerial
+            )
+        } ?: throw InstallException("Timeout transferring absolutePath")
+        return null
```
InstallSplitPackageRequest возвращает Unit. Нормально возвращать null?

Malinskiy replied:
Я бы просто `Unit` вернул

princeparadoxes replied:
В реализации этого метода в DdmlibAndroidDevice возвращается сообщение типа String. Поэтому в любом случае String стоит возвращать. 
Тут скорее вопрос в том, возвращать null или "Success"?

Commit:
Add return result on install split

Malinskiy committed:
fix(android): add tests for split apk installation

Malinskiy committed:
docs(android): update docs for split apk

Malinskiy committed:
fix(gradle-plugin): remove split apks from gradle plugin

These should be picked up automatically by plugin but the current AGP doesn't seem to expose these publicly

Malinskiy committed:
fix(android): remove verbose netty+vertx logging during tests

Commit:
Merge branch 'dynamic_feature_update' into dynamic_feature

# Conflicts:
#	vendor/vendor-android/adam/src/main/kotlin/com/malinskiy/marathon/android/adam/AdamAndroidDevice.kt

Commit:
Remove split apk from gradle extension

Commit:
Revert "Remove split apk from gradle extension"

This reverts commit 6778f10b4fe4d3fd2cf5ef9f3bfb116876870577.

Commit:
Add test for parse multi module testing
